api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2022.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// All fields in this package are required unless Explicitly marked optional
+// +kubebuilder:validation:Required
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// AdminNetworkPolicy is the Schema for the adminnetworkpolicies API
+type AdminNetworkPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of AdminNetworkPolicy.
+	Spec AdminNetworkPolicySpec `json:"spec,omitempty"`
+
+	// ANPStatus is the status to be reported by the implementation, this is not
+	// standardized in alpha and consumers should report what they see fit in
+	// relation to their AdminNetworkPolicy implementation
+	// +optional
+	Status AdminNetworkPolicyStatus `json:"status,omitempty"`
+}
+
+// AdminNetworkPolicyStatus defines the observed state of AdminNetworkPolicy
+type AdminNetworkPolicyStatus struct {
+	Conditions []metav1.Condition `json:"status"`
+}
+
+// AdminNetworkPolicySubject defines what objects the policy selects.
+// Exactly one of the `NamespaceSelector` or `NamespaceAndPodSelector` pointers
+// should be set.
+type AdminNetworkPolicySubject struct {
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector"`
+	// +optional
+	NamespaceAndPodSelector *NamespacedPodSubject `json:"namespaceAndPodSelector"`
+}
+
+// NamespacedPodSubject allows the user to select a given set of pod(s) in
+// selected namespace(s)
+type NamespacedPodSubject struct {
+	// This field follows standard label selector semantics; if present but empty,
+	// it selects all Namespaces.
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector"`
+
+	// Used to explicitly select pods within a namespace; if present but empty,
+	// it selects all Pods.
+	PodSelector *metav1.LabelSelector `json:"podSelector"`
+}
+
+// AdminNetworkPolicySpec defines the desired state of AdminNetworkPolicy
+type AdminNetworkPolicySpec struct {
+	// Priority is an int32 value bound to 0 - 1000, the lowest positive priority,
+	// "1" corresponds to the highest importance, while higher priorities have
+	// lower importance. An ANP with a priority of "0" will be evaluated after all
```

Comment:
I still think this is not a great approach.  

* 0 is less than 1 so it looks like it should be higher importance but actually it's a special case.  From this point of view, I think `nil` would be better.
* Since ANP allows for deny policies, order/priority matters so things get confusing if you have two policies at priority 0, one with allow and one with deny.  In Calico we use a deterministic tie breaker to make sure we at least do something consistent but it's always better if the user specifies the `order` field explicitly.

I think it'd be better to be explicit and use some sort of enum, for example, by splitting this field into two:
```
priorityClass: Override | Deferred
(or)           High | Low
(or)           AboveNetworkPolicy | BelowNetworkPolicy
(or)           OverrideNetworkPolicy | DeferToNetworkPolicy
priority: 0-1000
```

Alternatively, you could define that normal network policies have priority=1000 and then 1-999 and 1001-... would have a natural meaning.

Reply:
So after a bit of conversation, the general opinion was that it may be easier to have a separate object completely for the "baseline" use case, mainly due to the following reasons, 

- The "action=Pass" construct no longer makes sense when used with priority=0 or "baseline" rules 
- The interaction of rules (baseline, K8s Netpol, ANP) would be much easier to describe within the documentation and api itself 
- It really seems like it would be easier usability wise as well 

That work will be pushed to this PR shortly